Add 2x new blog posts and misc updates (#568)
hashintel/hash: sites/hashdev/src/_pages/blog/2_open-source.mdx

---
authorPhoto: blog/2_open-source/david.jpg
postPhoto: blog/2_open-source/OpenSource_Cover.jpg

> title: "Open-source at HASH"

Reviewer: ```suggestion
title: "Open source at HASH"
```

> subtitle: "How our approach to open-source informs what we do"

Reviewer: ```suggestion
subtitle: "How our approach to open source informs what we do"
```

author: "David Wilkinson"
jobTitle: "CEO at HASH"
date: "May 15th, 2022"
---

## Working in public

At HASH, we publish as much of our code as possible in public.

Depending on the code in question, it'll typically be released under one of three types of license, each of which has slightly different terms attached to it governing what you can and can't do with the underlying source:

> 1.  **Open-source and permissive licenses:** in essence these licenses give people the freedom to do pretty much whatever they like, including extend the code and keep it private (making this the most flexible category of license). The entirety of the [Block Protocol](https://github.com/blockprotocol/blockprotocol) is available under these terms (by way of the _MIT License_), alongside most of our blocks and connectors.

Reviewer: I _think_? It read a bit wrong at first, now I'm not sure.
```suggestion
1.  **Open source and permissive licenses:** in essence these licenses give people the freedom to do pretty much whatever they like, including extending the code and keeping it private (making this the most flexible category of license). The entirety of the [Block Protocol](https://github.com/blockprotocol/blockprotocol) is available under these terms (by way of the _MIT License_), alongside most of our blocks and connectors.
```

> 1.  **Open-source and copyleft licenses:** these again give people the freedom to use the code however they like, provided that if they extend or modify the code they contribute those changes back into the open-source as well under the same terms. Our [HASH](https://hash.ai/platform/hash) application is available under the _AGPL_ which guarantees this freedom.

Reviewer: ```suggestion
1.  **Open-source and copyleft licenses:** these again give people the freedom to use the code however they like, provided that if they extend or modify the code they contribute those changes back into the open source as well under the same terms. Our [HASH](https://hash.ai/platform/hash) application is available under the _AGPL_ which guarantees this freedom.
```

> 1.  **Fair-source licenses:** fair-source licenses gives people access to and freedom to use code directly, and the ability to self-host products without reliance on their creators (i.e. us). However fair-source licenses limit the ability of others to create competing commercial offerings based on fair-source code. [HASH Engine](https://hash.ai/platform/engine) is released under the _Elastic License (ELv2)_ which allows anyone to use, copy, distribute, make available, and prepare derivative works. However it restricts folks from using the software to provide a managed service to others (e.g. to provide a cloud simulation SaaS offering).

Reviewer: ```suggestion
1.  **Fair-source licenses:** fair-source licenses gives people access to and freedom to use code directly, and the ability to self-host products without reliance on their creators (i.e. us). However fair-source licenses limit the ability of others to create competing commercial offerings based on fair source code. [HASH Engine](https://hash.ai/platform/engine) is released under the _Elastic License (ELv2)_ which allows anyone to use, copy, distribute, make available, and prepare derivative works. However it restricts folks from using the software to provide a managed service to others (e.g. to provide a cloud simulation SaaS offering).
```

> ## How we think about open-source

Reviewer: ```suggestion
## How we think about open source
```

> First, **we contribute** original open-source software (OSS) into the ecosystem through our [HASH platform](https://hash.ai/platform/hash), its connectors, and our blocks. You can [find these](https://github.com/hashintel/hash/tree/main/packages/hash) on GitHub.

Reviewer: The link only goes to `hash` (since `blocks` are in a separate folder), so it's technically not pointing to "these"

Author: Changed to point to repo root

> Second, **we leverage** existing OSS within our platform to accelerate development of our products and to provide users with best-in-class experiences. Under the hood we rely on great technologies such as OpenSearch (Apache), Snowplow (Apache), Meltano (MIT), dbt (Apache), Great Expectations (Apache), and GitLab (MIT). In addition, we use dozens of other open-source packages within our extended codebase.

Reviewer: I'd consider swapping the integrations ones out for React, Apollo, NextJS, or some other big ones, in case we end up not using one or more of the ones listed

Author: Good call, thanks.

> Third, **we publish** a lot of the source code for the non-open-source components of our platform (such as [hEngine](https://hash.ai/platform/engine)) publicly under fair-code licenses like the Elastic License. This allows any HASH user to utilize our simulation technology within their organization for internal-use, free from restriction, provided they don’t offer this technology to others as a managed service themselves (as that would be competitive with our core business). In addition, the Elastic License forbids users from circumventing any license key functionality, or removing/obscuring any feature protected by those keys, as well as any included licenses, copyright and other legal notices.

Reviewer: ```suggestion
Third, **we publish** a lot of the source code for the non-open-source components of our platform (such as [hEngine](https://hash.ai/platform/engine)) publicly under fair-source licenses like the Elastic License. This allows any HASH user to utilize our simulation technology within their organization for internal use, free from restriction, provided they don’t offer this technology to others as a managed service themselves (as that would be competitive with our core business). In addition, the Elastic License forbids users from circumventing any license key functionality, or removing/obscuring any feature protected by those keys, as well as any included licenses, copyright and other legal notices.
```

> Finally, **we regularly evaluate the competitive landscape** to see if we can move private code into public repositories, and/or relicense existing public fair-code as open-source. When the environment changes, so do we. In keeping with [our mission](https://hash.ai/about/mission) to continuously reduce and eliminate information failures, our ultimate goal is to have a fully sustainable business with as large an open-source footprint as possible. We know we can’t even come close to conceiving all of the ways our technology might eventually be used by others, and it’s important to us to provide future innovators with as much freedom and flexibility as possible — without compromising our own ability to pursue HASH’s mission over the sustainable long-term.

Reviewer: ```suggestion
Finally, **we regularly evaluate the competitive landscape** to see if we can move private code into public repositories, and/or relicense existing public fair source as open source. When the environment changes, so do we. In keeping with [our mission](https://hash.ai/about/mission) to continuously reduce and eliminate information failures, our ultimate goal is to have a fully sustainable business with as large an open-source footprint as possible. We know we can’t even come close to conceiving all of the ways our technology might eventually be used by others, and it’s important to us to provide future innovators with as much freedom and flexibility as possible — without compromising our own ability to pursue HASH’s mission over the sustainable long-term.
```

Our open-source strategy stands in contrast to the approach of others within the structured-knowledge and simulation space, who argue that open-sourcing their software would reduce its quality (see: <a href="https://blog.wolfram.com/2019/04/02/why-wolfram-tech-isnt-open-source-a-dozen-reasons/" rel="nofollow" target="_blank">Why Wolfram Tech Isn’t Open Source—A Dozen Reasons</a>). We passionately but respectfully believe the opposite to be true.

## What Wolfram gets wrong

Wolfram are one of the oldest players in the structured knowledge space. Their mission, "to make it possible to compute whatever can be computed, whenever and wherever it is needed, and to make accessible the full frontiers of the computational universe" is highly aligned with our own.

> However, all twelve of Wolfram's stated criticisms of open-source are straw-men or outright falsifiable, as explored below.

Reviewer: ```suggestion
However, all twelve of Wolfram's stated criticisms of open source are straw-men or outright falsifiable, as explored below.
```

> 1.  **Strong leadership and design can still be provided within open-source.** Wolfram's first "argument" against open-source is that "a coherent vision requires centralized design" (1). This claim disregards the very many OSS projects with benevelolent dictators who impose design standards, constraints and their own ideas upon the evolution of their projects, much akin to how any company management might. Their second claim, in defense of Wolfram's being closed-source, is that "high-level languages need more design than low-level languages" (2) which it also follows is therefore not an argument against OSS either. "Crowd-sourced decisions can be bad for you" (5), as well as "Unified computation" and "Unified representation" requiring "unified design" (7 & 8) can be dismissed similarly, as can "Bad design is expensive" (12). _None of these are arguments against open-source so much as they are arguments for strong leadership, which remains eminently possible within an open-source environment._

Reviewer: ```suggestion
1.  **Strong leadership and design can still be provided within open source.** Wolfram's first "argument" against open source is that "a coherent vision requires centralized design" (1). This claim disregards the very many OSS projects with benevelolent dictators who impose design standards, constraints and their own ideas upon the evolution of their projects, much akin to how any company management might. Their second claim, in defense of Wolfram's being closed source, is that "high-level languages need more design than low-level languages" (2) which it also follows is therefore not an argument against OSS either. "Crowd-sourced decisions can be bad for you" (5), as well as "Unified computation" and "Unified representation" requiring "unified design" (7 & 8) can be dismissed similarly, as can "Bad design is expensive" (12). _None of these are arguments against open source so much as they are arguments for strong leadership, which remains eminently possible within an open-source environment._
```

> 1.  **Open-source communities are more diverse, not less.** Wolfram argue that "you need multidisciplinary teams to unify disparate fields" (3), a statement that is undoubtedly true. Yet claiming this as an argument _against_ open-source turns a blind eye to the diversity and breadth of expertise found within many OSS projects, including HASH, whose contributors come from multiple continents, have varied backgrounds, and in fact represent a much wider set of life-experiences and disciplines than a single company alone might hope to capture within its organizational structure.

Reviewer: ```suggestion
1.  **Open-source communities are more diverse, not less.** Wolfram argue that "you need multidisciplinary teams to unify disparate fields" (3), a statement that is undoubtedly true. Yet claiming this as an argument _against_ open source turns a blind eye to the diversity and breadth of expertise found within many OSS projects, including HASH, whose contributors come from multiple continents, have varied backgrounds, and in fact represent a much wider set of life-experiences and disciplines than a single company alone might hope to capture within its organizational structure.
```

1.  **Commercial entities can help open-source projects be more effective.** Wolfram argue that "hard cases and boring stuff need to get done too" (4), essentially implying that open-source maintainers and coders can't be trusted to knuckle down and focus on important maintainability, or boring background work required to make new technologies impactful and successful. Not only is this patronizing to the thousands of open-source developers who care passionately about the products they build, but it ignores the possibility of commercial entities advancing OSS projects and acting with similar motivations to any other private company. "Our developers work for you, not just themselves" (6) relies upon the same misrepresentation, while "Paid software offers an open quid pro quo" (10) mischaracterizes and ignores the various business models that allow OSS to be both free at the point of access, and profitable (e.g. through the provision of paid complements, such as the famous support services provided by _Red Hat_ which justified their $34 billion sale to _IBM_ in 2019).

> 1.  **Huge technological advances occur in the open.** Wolfram's claim that "Open-source doesn’t bring major tech innovation to market" (9) can be disproven by a thousand major projects, and the entirety of the modern web. We won't list the many projects here that prove this statement's desperate vacuity.

Reviewer: ```suggestion
1.  **Huge technological advances occur in the open.** Wolfram's claim that "Open source doesn’t bring major tech innovation to market" (9) can be disproven by a thousand major projects, and the entirety of the modern web. We won't list the many projects here that prove this statement's desperate vacuity.
```

> 1.  **There are many long-term financing options and sustainable business models available to open-source projects.** According to Wolfram, "it takes steady income to sustain long-term R&D" (11), which is generally true (unless you happen to be sat on a large nest-egg already). However, many OSS projects are able to command a stable income, making this again _not an argument against open-source at all_. In fact, there are various OSS companies that are far more profitable than Wolfram. Since the advent of debt in approximately 3500 BC it's been possible to operate without monetizing upfront -- in promise or expectation of future cash-flows -- and modern financing in the form of venture capital is just one form of this that has led to an explosion in OSS companies. To ignore the last five millenia of human experience and financial innovation is disingenuous.

Reviewer: ```suggestion
1.  **There are many long-term financing options and sustainable business models available to open-source projects.** According to Wolfram, "it takes steady income to sustain long-term R&D" (11), which is generally true (unless you happen to be sat on a large nest-egg already). However, many OSS projects are able to command a stable income, making this again _not an argument against open source at all_. In fact, there are various OSS companies that are far more profitable than Wolfram. Since the advent of debt in approximately 3500 BC it's been possible to operate without monetizing upfront -- in promise or expectation of future cash-flows -- and modern financing in the form of venture capital is just one form of this that has led to an explosion in OSS companies. To ignore the last five millenia of human experience and financial innovation is disingenuous.
```

The likes of Wolfram find themselves constrained by path-dependency, struggling to justify historical choices and lumbered with a conseqeuntly outdated business model. Changing their model now would risk cannibalizing their existing paying user-base. We, in constrast, are relatively unencumbered, and able to make better product design and institutional decisions.

> As Wolfram note, great design requires effective decision-making at the top, and this at the heart of everything we do at HASH. That is not, however, mutually-exclusive with open-source - and the very best design benefits from a wide-range of inputs. Open-source is a fundamentally better approach to robust, reliable, and effective software development. That is what we're all about.

Reviewer: ```suggestion
As Wolfram note, great design requires effective decision-making at the top, and this at the heart of everything we do at HASH. That is not, however, mutually-exclusive with open source - and the very best design benefits from a wide-range of inputs. Open-source is a fundamentally better approach to robust, reliable, and effective software development. That is what we're all about.
```
